File: src/zerotts/codec.py
"""onnxruntime-only MOSS-Audio-Tokenizer-Nano **decoder** — codes -> waveform.

Vendored, not downloaded. The decoder graphs ship inside the ZeroTTS weights
repo (``onnx/codec/``) so a ZeroTTS install has no runtime dependency on any
third-party model repo staying up or unchanged.

Decoder only. The upstream export also has an encoder graph (waveform -> codes),
used solely to turn reference audio into voice latents. That is voice cloning,
which this release does not do (see zerotts.voices), so the encoder is neither
shipped nor wrapped — it would be ~45MB of weights nothing here can call.

Conventions worth knowing before touching this:
  * native sample rate 48 kHz; the codec is stereo internally, the public
    interface is mono (decode averages the two channels).
  * this export uses (batch, T, K) — TIME-major, codebook-LAST — for
    ``audio_codes``. ZeroTTS's AR loop produces (B, K, T), so decode transposes.
  * every integer tensor here is **int32**, not int64. This is verified against
    the graphs; do not "fix" it to int64 out of PyTorch habit.

Credit: MOSS-Audio-Tokenizer-Nano by the OpenMOSS team, Apache-2.0. See
https://github.com/OpenMOSS/MOSS-Audio-Tokenizer and the NOTICE file.
"""

from __future__ import annotations

import json
from pathlib import Path

import numpy as np


class MossCodecDecoder:
    """Args:
        codec_dir: directory holding the decoder graphs and
            ``codec_browser_onnx_meta.json``.
        providers: onnxruntime execution providers.
        intra_op_num_threads: per-session thread count.
    """

    def __init__(
        self,
        codec_dir: str | Path,
        providers: list[str] | None = None,
        intra_op_num_threads: int = 4,
    ):
        import onnxruntime as ort

        codec_dir = Path(codec_dir)
        meta_path = codec_dir / "codec_browser_onnx_meta.json"
        if not meta_path.exists():
            raise FileNotFoundError(
                f"{codec_dir} has no codec_browser_onnx_meta.json — the vendored "
                f"codec is missing from this model directory.")
        meta = json.loads(meta_path.read_text())
        self._meta = meta

        cfg = meta["codec_config"]
        self.sample_rate = int(cfg["sample_rate"])
        self.num_channels = int(cfg["channels"])
        self.frame_size = int(cfg["downsample_rate"])
        self.frame_rate = self.sample_rate / self.frame_size
        self.num_codebooks = int(cfg["num_quantizers"])

        sess_options = ort.SessionOptions()
        sess_options.graph_optimization_level = ort.GraphOptimizationLevel.ORT_ENABLE_ALL
        sess_options.intra_op_num_threads = intra_op_num_threads
        sess_options.inter_op_num_threads = 1
        resolved = providers or ["CPUExecutionProvider"]

        def _session(key: str):
            return ort.InferenceSession(
                str(codec_dir / meta["files"][key]),
                sess_options=sess_options,
                providers=resolved,
            )

        self._decode_full_sess = _session("decode_full")
        self._decode_step_sess = _session("decode_step")

    def decode(self, codes_bkt: np.ndarray) -> np.ndarray:
        """(B, K, T) int codes -> (B, T_audio) float32 mono at sample_rate."""
        codes = np.asarray(codes_bkt)
        if codes.ndim == 2:  # (K, T) -> (1, K, T)
            codes = codes[None, :, :]
        codes_btk = codes.transpose(0, 2, 1).astype(np.int32)  # -> (B, T, K)
        lengths = np.array([codes_btk.shape[1]], dtype=np.int32)
        audio, audio_lengths = self._decode_full_sess.run(
            None, {"audio_codes": codes_btk, "audio_code_lengths": lengths}
        )
        n = int(audio_lengths.reshape(-1)[0])
        return audio[:, :, :n].mean(axis=1).astype(np.float32)

    def streaming_decoder(self) -> MossStreamingDecoder:
        """Open a stateful streaming decoder (keeps the causal decoder KV cache
        across chunks). Call decode_chunk per chunk, then close."""
        return MossStreamingDecoder(self)


class MossStreamingDecoder:
    """KV-cached streaming decode over decode_step.onnx, driven by the state
    layout ``codec_browser_onnx_meta.json``'s "streaming_decode" section
    describes: per-decoder transformer offsets plus per-layer attention caches
    (key/value/position ring buffers). Use via
    ``MossCodecDecoder.streaming_decoder()``."""

    def __init__(self, codec: MossCodecDecoder):
        self._codec = codec
        self._session = codec._decode_step_sess
        streaming = codec._meta.get("streaming_decode", {})
        self._transformer_specs = list(streaming.get("transformer_offsets", []))
        self._attention_specs = list(streaming.get("attention_caches", []))
        self._output_names = [o.name for o in self._session.get_outputs()]
        self._state: dict = {}
        self._reset_state()

    def _reset_state(self) -> None:
        self._state = {}
        for spec in self._transformer_specs:
            self._state[str(spec["input_name"])] = np.zeros(tuple(spec["shape"]), dtype=np.int32)
        for spec in self._attention_specs:
            self._state[str(spec["offset_input_name"])] = np.zeros(
                tuple(spec["offset_shape"]), dtype=np.int32)
            self._state[str(spec["cached_keys_input_name"])] = np.zeros(
                tuple(spec["cache_shape"]), dtype=np.float32)
            self._state[str(spec["cached_values_input_name"])] = np.zeros(
                tuple(spec["cache_shape"]), dtype=np.float32)
            # -1, not 0: position 0 is a real position, so a zero-filled ring
            # buffer would read as "every slot holds frame 0".
            self._state[str(spec["cached_positions_input_name"])] = np.full(
                tuple(spec["positions_shape"]), -1, dtype=np.int32)

    def decode_chunk(self, codes_bkt: np.ndarray) -> np.ndarray:
        """(1, K, n) int codes -> (1, chunk_samples) float32 mono."""
        codes = np.asarray(codes_bkt)
        if codes.ndim == 2:
            codes = codes[None, :, :]
        codes_btk = codes.transpose(0, 2, 1).astype(np.int32)  # -> (1, n, K)
        feeds = {
            "audio_codes": codes_btk,
            "audio_code_lengths": np.array([codes_btk.shape[1]], dtype=np.int32),
            **self._state,
        }
        outputs = self._session.run(None, feeds)
        named = dict(zip(self._output_names, outputs))
        for spec in self._transformer_specs:
            self._state[str(spec["input_name"])] = named[str(spec["output_name"])]
        for spec in self._attention_specs:
            for key in ("offset", "cached_keys", "cached_values", "cached_positions"):
                self._state[str(spec[f"{key}_input_name"])] = named[str(spec[f"{key}_output_name"])]
        n = int(named["audio_lengths"].reshape(-1)[0])
        return named["audio"][:, :, :n].mean(axis=1).astype(np.float32)

    def close(self) -> None:
        self._reset_state()

File: js/src/codec.ts
/**
 * MOSS-Audio-Tokenizer-Nano decoder — codes to waveform.
 *
 * Port of zerotts/codec.py. Decoder only; the encoder graph is not shipped
 * (it exists for voice cloning, which this release does not do).
 *
 * Three things differ from the TTS graphs and cause most porting bugs:
 *   - every integer tensor here is int32, NOT int64
 *   - the code axis order is (B, T, K) — time-major, codebook-last — while the
 *     TTS loop produces (B, K, T), so decode transposes
 *   - the streaming position ring buffer initializes to -1, not 0: position 0 is
 *     a real position, so a zero-filled buffer reads as "every slot holds
 *     frame 0"
 *
 * Credit: OpenMOSS team, Apache-2.0. See NOTICE.
 */

import * as ort from 'onnxruntime-web';

interface TransformerOffsetSpec {
  input_name: string;
  output_name: string;
  shape: number[];
}

interface AttentionCacheSpec {
  offset_input_name: string;
  offset_output_name: string;
  cached_keys_input_name: string;
  cached_keys_output_name: string;
  cached_values_input_name: string;
  cached_values_output_name: string;
  cached_positions_input_name: string;
  cached_positions_output_name: string;
  offset_shape: number[];
  cache_shape: number[];
  positions_shape: number[];
}

export interface CodecMeta {
  files: Record<string, string>;
  codec_config: {
    sample_rate: number;
    channels: number;
    downsample_rate: number;
    num_quantizers: number;
  };
  streaming_decode?: {
    transformer_offsets: TransformerOffsetSpec[];
    attention_caches: AttentionCacheSpec[];
  };
}

const prod = (dims: number[]) => dims.reduce((a, b) => a * b, 1);

/** (B, K, T) -> (B, T, K) int32, the layout the codec graphs expect. */
function transposeToBTK(codes: Int32Array, B: number, K: number, T: number): Int32Array {
  const out = new Int32Array(B * T * K);
  for (let b = 0; b < B; b++) {
    for (let k = 0; k < K; k++) {
      for (let t = 0; t < T; t++) {
        out[b * T * K + t * K + k] = codes[b * K * T + k * T + t];
      }
    }
  }
  return out;
}

/** (B, C, T) float32 -> (B, T) mono by averaging channels. */
function toMono(data: Float32Array, B: number, C: number, T: number, n: number): Float32Array {
  const out = new Float32Array(n);
  for (let t = 0; t < n; t++) {
    let sum = 0;
    for (let c = 0; c < C; c++) sum += data[c * T + t];
    out[t] = sum / C;
  }
  return out;
}

export class MossCodecDecoder {
  readonly sampleRate: number;
  readonly numChannels: number;
  readonly numCodebooks: number;
  readonly frameSize: number;

  private constructor(
    private decodeFull: ort.InferenceSession,
    private decodeStep: ort.InferenceSession,
    private meta: CodecMeta,
  ) {
    const cfg = meta.codec_config;
    this.sampleRate = cfg.sample_rate;
    this.numChannels = cfg.channels;
    this.numCodebooks = cfg.num_quantizers;
    this.frameSize = cfg.downsample_rate;
  }

  static async create(
    meta: CodecMeta,
    graphs: { decodeFull: ArrayBuffer; decodeStep: ArrayBuffer },
    options: ort.InferenceSession.SessionOptions,
  ): Promise<MossCodecDecoder> {
    const [full, step] = await Promise.all([
      ort.InferenceSession.create(graphs.decodeFull, options),
      ort.InferenceSession.create(graphs.decodeStep, options),
    ]);
    return new MossCodecDecoder(full, step, meta);
  }

  /** (K, T) codes -> mono float32 at sampleRate. One call for the utterance. */
  async decode(codesKT: Int32Array, K: number, T: number): Promise<Float32Array> {
    const btk = transposeToBTK(codesKT, 1, K, T);
    const outputs = await this.decodeFull.run({
      audio_codes: new ort.Tensor('int32', btk, [1, T, K]),
      audio_code_lengths: new ort.Tensor('int32', Int32Array.from([T]), [1]),
    });
    const audio = outputs.audio as ort.Tensor;
    const lengths = outputs.audio_lengths as ort.Tensor;
    const n = Number((lengths.data as Int32Array)[0]);
    const [, C, Tfull] = audio.dims as number[];
    return toMono(audio.data as Float32Array, 1, C, Tfull, n);
  }

  streamingDecoder(): MossStreamingDecoder {
    return new MossStreamingDecoder(this.decodeStep, this.meta);
  }
}

export class MossStreamingDecoder {
  private state = new Map<string, ort.Tensor>();
  private transformerSpecs: TransformerOffsetSpec[];
  private attentionSpecs: AttentionCacheSpec[];

  constructor(private session: ort.InferenceSession, meta: CodecMeta) {
    const streaming = meta.streaming_decode ?? { transformer_offsets: [], attention_caches: [] };
    this.transformerSpecs = streaming.transformer_offsets;
    this.attentionSpecs = streaming.attention_caches;
    this.reset();
  }

  reset(): void {
    this.state.clear();
    for (const spec of this.transformerSpecs) {
      this.state.set(spec.input_name,
        new ort.Tensor('int32', new Int32Array(prod(spec.shape)), spec.shape));
    }
    for (const spec of this.attentionSpecs) {
      this.state.set(spec.offset_input_name,
        new ort.Tensor('int32', new Int32Array(prod(spec.offset_shape)), spec.offset_shape));
      this.state.set(spec.cached_keys_input_name,
        new ort.Tensor('float32', new Float32Array(prod(spec.cache_shape)), spec.cache_shape));
      this.state.set(spec.cached_values_input_name,
        new ort.Tensor('float32', new Float32Array(prod(spec.cache_shape)), spec.cache_shape));
      // -1, not 0 — see the file header.
      const positions = new Int32Array(prod(spec.positions_shape)).fill(-1);
      this.state.set(spec.cached_positions_input_name,
        new ort.Tensor('int32', positions, spec.positions_shape));
    }
  }

  /** (K, frames) codes for this chunk -> mono float32. */
  async decodeChunk(codesKT: Int32Array, K: number, frames: number): Promise<Float32Array> {
    const btk = transposeToBTK(codesKT, 1, K, frames);
    const feeds: Record<string, ort.Tensor> = {
      audio_codes: new ort.Tensor('int32', btk, [1, frames, K]),
      audio_code_lengths: new ort.Tensor('int32', Int32Array.from([frames]), [1]),
    };
    for (const [k, v] of this.state) feeds[k] = v;

    const outputs = await this.session.run(feeds);

    for (const spec of this.transformerSpecs) {
      this.state.set(spec.input_name, outputs[spec.output_name] as ort.Tensor);
    }
    for (const spec of this.attentionSpecs) {
      this.state.set(spec.offset_input_name, outputs[spec.offset_output_name] as ort.Tensor);
      this.state.set(spec.cached_keys_input_name, outputs[spec.cached_keys_output_name] as ort.Tensor);
      this.state.set(spec.cached_values_input_name, outputs[spec.cached_values_output_name] as ort.Tensor);
      this.state.set(spec.cached_positions_input_name, outputs[spec.cached_positions_output_name] as ort.Tensor);
    }

    const audio = outputs.audio as ort.Tensor;
    const lengths = (outputs.audio_lengths as ort.Tensor).data as Int32Array;
    const n = Number(lengths[0]);
    const [, C, T] = audio.dims as number[];
    return toMono(audio.data as Float32Array, 1, C, T, n);
  }
}
